.storybook/grid/GridPanel.tsx

```diff
@@ -0,0 +1,115 @@
+import React, { useEffect, useState, useRef } from 'react';
+import { AddonPanel } from '@storybook/components';
+import { useChannel } from '@storybook/api';
+import { styles } from './gridStyles';
+import { INITIAL_IMAGE_EVENT, IMAGE_SELECTED_EVENT } from './withGrid';
+import { STORY_CHANGED } from '@storybook/core-events';
+import { IframePostMessageService } from '@guardian/grid-client';
+import { Reporter } from '@guardian/grid-client/lib/utils';
+
+const GRID_URL =
+    process.env.NODE_ENV === 'development'
+        ? 'https://media.test.dev-gutools.co.uk'
+        : 'https://media.gutools.co.uk';
+
+const isValidMessage = (data) => data?.crop?.data && data?.image?.data;
+
+export const GridPanel = ({ api, active }) => {
+    const emit = useChannel({});
+    const [isModalOpen, setIsModalOpen] = useState(false);
+    const [selectedImage, setSelectedImage] = useState();
+    const activeButton = useRef(null);
+
+    const onMessage = (event) => {
+        if (event.origin !== GRID_URL) {
+            return;
+        }
+
+        const postMessageService = new IframePostMessageService(event, Reporter.default);
+
+        if (!postMessageService.isValid) {
+            return;
+        }
+
+        const cropAssets = event.data.crop.data.assets;
+        const cropUrl = cropAssets[cropAssets.length - 1]?.secureUrl;
+        setSelectedImage(cropUrl);
+        emit(IMAGE_SELECTED_EVENT, cropUrl);
+
+        setIsModalOpen(false);
+    };
+
+    useEffect(() => {
+        if (isModalOpen) {
+            window.addEventListener('message', onMessage, false);
+        } else {
+            window.removeEventListener('message', onMessage, false);
+        }
+
+        return () => window.removeEventListener('message', onMessage, false);
+    }, [isModalOpen]);
```

Comment:
I can't see how the inner scope `useEffect` will know the name of the `isModelOpen` variable, if it's only passed in as an array rather than an object in its second argument...

Reply:
Hey good question. The scoping here is that the `isModelOpen` const is defined outside the first argument to `useEffect` (the callback) so it closes over it. The second (optional) argument to `useEffect` is the list of dependencies, so at the end of the current render cycle the effect will be called only if `isModalOpen` has changed.